Feature Dependency and Impact Analysis: Dependency Modeling & Persistence (#82)

DmitriyChernyakSetronica opened the pull request:
# Objective
Define the data structure for feature dependencies and ensure they are correctly stored in the database with referential integrity.

# Requirements
- **Feature Dependency Entity**: Create a database entity or model with the fields:
    - id
    - feature (many-to-one to dependent feature using feature.code)
    - dependsOnFeature (many-to-one to dependency feature using feature.code)
    - dependencyType (hard/soft/optional)
    - createdAt
    - notes

- **Relationships**: Implement the bidirectional one-to-many relationship allowing any feature to have zero or more dependencies.

- **Database Schema**: Ensure the database schema is updated to store dependencies and enforces referential integrity (e.g., a dependency cannot be created for a non-existent featureCode).

- **CRUD Repository**: 
    - Implement `FeatureDependencyRepository` using Spring Data JPA.
    - Repository must support nested property queries, e.g.:
        - `findByFeature_Code`
        - `findByDependsOnFeature_Code`
        - `findByFeature_CodeAndDependsOnFeature_Code`


# Test Coverage
- Unit tests for the FeatureDependency entity/model.
- Integration tests for database persistence (create, read, update, delete).
- Tests to ensure database constraints (like foreign keys) prevent invalid data entry and enforce unique constraints preventing duplicate dependencies and self-dependencies.

# Acceptance Criteria
- The FeatureDependency entity is correctly defined in the codebase and migrated to the database.
- Dependencies can be successfully saved to, and retrieved from, the database.
- The system prevents the creation of dependencies that reference non-existent features.

FAIL_TO_PASS: com.sivalabs.ft.features.domain.FeatureDependencyRepositoryTest.java, com.sivalabs.ft.features.domain.FeatureDependencyIntegrityTest.java, com.sivalabs.ft.features.domain.entities.FeatureDependencyTest.java

volkov commented on src/test/java/com/sivalabs/ft/features/domain/entities/FeatureDependencyTest.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    @Test
+    void shouldHandleAllDependencyTypes() {
```
This test probably could be parametrised.

volkov commented on src/main/java/com/sivalabs/ft/features/domain/FeatureDependencyRepository.java:
```diff
@@ -0,0 +1,36 @@
+
+    List<FeatureDependency> findByFeature_Code(String featureCode);
+
+    List<FeatureDependency> findByDependsOnFeature_Code(String dependsOnFeatureCode);
```
findByFeature_Code is specified in description.

But all other methods are not, and used in tests.

This should be fixed.

volkov commented on src/main/java/com/sivalabs/ft/features/domain/entities/Feature.java:
```diff
@@ -1,10 +1,13 @@
 package com.sivalabs.ft.features.domain.entities;
 
+import com.fasterxml.jackson.annotation.JsonManagedReference;
```
I see that there is FeatureDTO - so I dont think that jackson annotations should be here. Lets remove them.

At least they do not affects tests.

DmitriyChernyakSetronica committed:
Feature Dependency and Impact Analysis: Dependency Modeling & Persistence #82
Objective
Define the data structure for feature dependencies and ensure they are correctly stored in the database with referential integrity.

Requirements
Feature Dependency Entity: Create a database entity or model with the fields:

id
feature (many-to-one to dependent feature using feature.code)
dependsOnFeature (many-to-one to dependency feature using feature.code)
dependencyType (hard/soft/optional)
createdAt
notes
Relationships: Implement the bidirectional one-to-many relationship allowing any feature to have zero or more dependencies.

Database Schema: Ensure the database schema is updated to store dependencies and enforces referential integrity (e.g., a dependency cannot be created for a non-existent featureCode).

CRUD Repository:

Implement FeatureDependencyRepository using Spring Data JPA.
Repository must support nested property queries, e.g.:
findByFeature_Code
findByDependsOnFeature_Code
findByFeature_CodeAndDependsOnFeature_Code
Test Coverage
Unit tests for the FeatureDependency entity/model.
Integration tests for database persistence (create, read, update, delete).
Tests to ensure database constraints (like foreign keys) prevent invalid data entry and enforce unique constraints preventing duplicate dependencies and self-dependencies.
Acceptance Criteria
The FeatureDependency entity is correctly defined in the codebase and migrated to the database.
Dependencies can be successfully saved to, and retrieved from, the database.
The system prevents the creation of dependencies that reference non-existent features.

FAIL_TO_PASS: com.sivalabs.ft.features.domain.FeatureDependencyRepositoryTest.java, com.sivalabs.ft.features.domain.FeatureDependencyIntegrityTest.java, com.sivalabs.ft.features.domain.entities.FeatureDependencyTest.java

DmitriyChernyakSetronica commented on src/test/java/com/sivalabs/ft/features/domain/FeatureDependencyIntegrityTest.java:
```diff
@@ -0,0 +1,351 @@
+
+    @Test
+    @Rollback
+    void shouldCascadeDeleteWhenFeatureIsDeleted() {
```
fixed by agent with new name

DmitriyChernyakSetronica committed:
Feature Dependency and Impact Analysis: Dependency Management API #83 (#78)

**Objective** Implement the core business logic and API endpoints for creating, updating, and deleting feature dependencies. ⚙️

**Requirements**

* **CRUD Logic:** Develop services to add, modify (e.g., change type or notes), and remove dependencies between features.
* **Data Handling:** Ensure that dependency types and notes are correctly processed and saved.

**API Endpoints**

* `POST /features/{featureCode}/dependencies`: Create a new dependency for a feature.
* `PUT /features/{featureCode}/dependencies/{dependsOnFeatureCode}`: Update an existing dependency (e.g., to change its type or notes).
* `DELETE /features/{featureCode}/dependencies/{dependsOnFeatureCode}`: Remove a dependency.
* Request payload for dependency creation with JSON structure:
        `{
         "dependsOnFeatureCode": "FEATURE-123", // Required: Target feature code that this feature depends on
         "dependencyType": "HARD|SOFT|OPTIONAL", // Required: Type of dependency relationship
         "notes": "Optional description" // Optional: Additional notes about the dependency
         }`
* Request payload for dependency updates with JSON structure:
        `{
         "dependencyType": "HARD|SOFT|OPTIONAL", // Required: Updated type of dependency relationship
         "notes": "Updated description" // Optional: Updated notes about the dependency (null to clear)
         }`

**Security**

* Implement authorization checks to ensure only authorized users can create, modify, or delete dependencies.

**Admin & Logging**

* Log all dependency modifications (create, update, delete) for audit and traceability.

**Test Coverage**

* Unit and integration tests for the dependency management services.
* API contract tests for the `POST`, `PUT`, and `DELETE` endpoints.
* Security tests to verify that unauthorized users cannot modify dependencies.

**Acceptance Criteria**

* Authorized users can successfully create, update, and remove dependencies via the API.
* All modification actions are logged.
* Unauthorized API calls are rejected with an appropriate error status.

FAIL_TO_PASS: FeatureControllerApiContractTests.java, FeatureControllerIntegrationTests.java, FeatureControllerTests.java
